### PR TITLE
MM-55192: updated moment to use device time zone

### PR DESCRIPTION
**Pull request by srisri332**
#### Summary

Moment JS is handling daylight savings for people who live in countries like the US, but it isn't handling when the time shift is in European countries. Instead of turning back the clock on 27th October 2024 by 1 hr., it is somehow adding an hour to the current time, which effectively makes it +2 hours forward, due to which application freezes. This is easily fixed by explicitly asking Moment JS to use the device time zone.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-55192
https://github.com/mattermost/mattermost/issues/25352

#### Release Note

```release-note
NA
```

### CODE REVIEW

**Comment by srisri332** [comment]
@hmhealey , please review the following PR.

**Review by devinbinnie — APPROVED** [review]
LGTM! Thanks @srisri332!

**Comment by lindy65** [comment]
Hi @srisri332 - I'm testing on the spinwick and am not able to select Sunday the 27th of October at all.  When I select this date, the date picker automatically selects Saturday the 26th of October. Please take a look? Thanks!

cc// @hmhealey

**Comment by srisri332** [comment]
I just checked this, and unfortunately for some reason the date on the calendar is coming up a day before the selected day. If we're selecting the 27th,  26th is reflected in the calendar. If we select 26th then 25th is shown, so on. I'll take a look why.

**Comment by lindy65** [comment]
These are my repro steps @srisri332 

- Make sure the time in you system is set to German time. Which is UTC + 1:00 Amesterdam, Berlin .... in windows.
- Login to the webapp as any user, go to profile, select set a custom status. 
- There, set any status, but in the "Clear After" dropdown, select set custom date and time. 
- Now, in the custom date, from the calendar drop down, select the last Sunday of October. Which would be 27-10-2024.
- Try selecting the time now between 1:00-2:00 am. The app would hang and not allow you to perform any operation.

File: webapp/channels/src/packages/mattermost-redux/src/actions/timezone.ts [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
+
 import {getCurrentTimezoneFull} from 'mattermost-redux/selectors/entities/timezone';
 import {getCurrentUser} from 'mattermost-redux/selectors/entities/users';
 import type {ActionFuncAsync} from 'mattermost-redux/types/actions';
 
 import {updateMe} from './users';
 
+
 export function autoUpdateTimezone(deviceTimezone: string): ActionFuncAsync {
+    moment.tz.setDefault(deviceTimezone);
```
Comment (devinbinnie) [inline webapp/channels/src/packages/mattermost-redux/src/actions/timezone.ts:14]: I think having this call outside of the Redux action isn't the best idea. @hmhealey is this okay?

Comment (srisri332) [reply webapp/channels/src/packages/mattermost-redux/src/actions/timezone.ts:14]: I agree with this, however, with my previous change, somehow it is setting the wrong date leading to unexpected functionality discussed above. I just made this change to check if this helps. In any way. I'm somewhat concluded  that setting the default timezone is not the way to fix this, at least here and I've to look for a way to do this via a different method.

**Comment by srisri332** [comment]
Hey @hmhealey , yes I wanted to find another way to fix this. I'm unsure how but I'll try to find a way. Is there any chance I can keep this PR open so that when I push some changes I'd have the spinwick instance for testing? Also, the spinwick build deployment keeps failing. I would appreciate it if someone helped me with that.

File: webapp/channels/src/packages/mattermost-redux/src/actions/timezone.ts (hunk 1 above)
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/actions/timezone.ts]: Not that it matters here any more, but the difference between calling stuff inside the returned function vs outside is minor from what I can tell. Calling it outside makes it happen before Redux middleware has processed the action and calling it inside has it happen while the Redux middleware is doing its thing. Typically, we want to call it inside just for consistency's sake.

**Comment by hmhealey** [comment]
Yeah, we can definitely leave this open. I'm not sure why it's still failing since those failures happened during the week or two where the logging for them was broken. I'll see if pushing a new version actually tells us what the heck is causing problems now

/update-branch

**Comment by hmhealey** [comment]
@srisri332 We're still trying to figure out what's wrong with the test server here. We've at least fixed the original issue (the EETest label wasn't applied to build enterprise edition), but now it's having other issues on the MM cloud side, so I don't know when we'll get it working. If you can, it'd probably be best to work off your local environment if you have one set up and if you can there

**Comment by srisri332** [comment]
Hey @hmhealey , I've been trying to reproduce the issue in my local system but without any success. Here the issue means, with the fix applied, the calendar reflecting the previous date instead of the selected one. But yes, I'm still trying to find if there is any other way to fix this.

**Comment by srisri332** [comment]
@lindy65 , i reverted the changes for the fix. I'm assuming that the new spinwick instance wouldn't have the changes. Please correct me if I'm wrong. If it doesn't have the changes, I'm not sure why the date picker is behaving oddly. Could it be because of something else?
Cc: @hmhealey

**Comment by lindy65** [comment]
Thanks @srisri332 - I'm going to update the branch and see if that reverts your changes.  If not, then I'm not sure what the issue could be

**Comment by hmhealey** [comment]
Pulling this back to a draft PR while we figure out an alternative solution for this

**Comment by lindy65** [comment]
Removed `QA-review` label until PR is ready for QA

**Comment by hmhealey** [comment]
Hi @srisri332. Are you still working on that alternate solution for this? If not, that's totally fine, but I'd like to reopen this ticket to others who might want to work on it.

Also, related to the original fix for this, we've had https://github.com/mattermost/mattermost/pull/26957 merged recently which makes it so that `autoUpdateTimezone` is dispatched properly whenever the time zone changes. I'm not sure if that's related to this bug though

**Comment by hmhealey** [comment]
Closing this to open up the ticket to others

**Pull request closed without merging**